libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -1645,4 +1690,51 @@ bool AP_UAVCAN::led_write(uint8_t led_index, uint8_t red, uint8_t green, uint8_t
     return true;
 }
 
+AP_UAVCAN::EscStatus_Data* AP_UAVCAN::escstatus_getptrto_data(uint8_t id)
+{
+    // I think the esc_index are continues, by how ArduPilot works
+    // so we could just directly jump with id into the data list, return &_escstatus.data[id], with id<8 overflow protection of course
+
+    // check if id is already in list, and if it is take it
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == id) {
+            return &_escstatus.data[i];
+        }
+    }
+
+    // if id is not yet in list, find the first free spot, and take that
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == UINT8_MAX) {
+            _escstatus.id[i] = id;
+            return &_escstatus.data[i];
+        }
+    }
+
+    return nullptr;
+}
+
+void AP_UAVCAN::escstatus_update_data(uint8_t id)
+{
+    // only 8 LOG_ESC1_MSG are defined, see /libraries/DataFlash/LogStructure.h
+    // technically, it could happen that the esc_index is not continues, and one would need a better handling
+    // however, I think, ArduPilot implicitly enforces continues esc_index, so should be no problem
```

Comment:
The comment is from author, so needs to be rephrased and found out exactly if there will be a problem or not.